v3/src/models/data/data-set.ts

```diff
@@ -483,83 +464,105 @@ export const DataSet = V2Model.named("DataSet").props({
-            const allAttrCount = self.attributes.length
-            const collectionAttrCount = self.collections
-                                          .reduce((sum, collection) => sum += collection.attributes.length, 0)
-            if (collectionAttrCount >= allAttrCount) {
-              result.removedCollectionId = self.ungrouped.id
-              self.collections.splice(self.collections.length - 1, 1)
-            }
-          }
-          this.invalidateCollectionGroups()
-        }
-        return result
-      },
-      // if beforeCollectionId is not specified, new collection is last (child-most)
-      moveAttributeToNewCollection(attributeId: string, beforeCollectionId?: string) {
-        const attribute = self.getAttribute(attributeId)
-        if (attribute) {
-          const name = pluralize(attribute.name)
-          const collection = CollectionModel.create({ name })
-          this.addCollection(collection, beforeCollectionId)
-          this.setCollectionForAttribute(attributeId, { collection: collection.id })
-          return collection
-        }
-      }
     }
   }
 })
+.actions(self => ({
+  moveAttribute(attributeID: string, options?: IMoveAttributeCollectionOptions): IAttributeChangeResult {
+    let removedCollectionId: string | undefined
+    const attribute = self.getAttribute(attributeID)
+    const srcCollection = self.getCollectionForAttribute(attributeID)
+    const dstCollection = options?.before
+                            ? self.getCollectionForAttribute(options.before)
+                            : options?.after
+                              ? self.getCollectionForAttribute(options.after)
+                              : options?.collection
+                                ? self.getCollection(options.collection)
+                                : undefined
+    if (!attribute || !srcCollection) return {}
+    if (!dstCollection || srcCollection === dstCollection) {
+      srcCollection?.moveAttribute(attributeID, options)
+    }
+    else {
+      if (attribute.hasFormula) {
+        // If the attribute has a formula, we need to reset all the calculated values to blank values so that they
+        // are not taken into account while calculating case grouping. After the grouping is done, the formula will
+        // be re-evaluated, and the values will be updated to the correct values again.
+        attribute.clearValues()
+      }
+      if (srcCollection.getAttribute(attributeID) && srcCollection.attributes.length === 1) {
+        removedCollectionId = srcCollection.id
+        self.removeCollection(srcCollection)
+      }
+      else {
+        srcCollection.removeAttribute(attributeID)
+      }
+      dstCollection.addAttribute(attribute, options)
+      // update grouping
+      self.invalidateCollectionGroups()
+    }
+    return { removedCollectionId }
+  }
+}))
+.actions(self => ({
+  // if beforeCollectionId is not specified, new collection is last (child-most)
```

Comment:
Is this comment accurate? Looking at `addCollection`, it looks like the default is to add left of the child-most collection. For what it's worth, I think adding either a child-most or parent-most collection would be better than left of the child-most collection.

Reply:
I experimented with changing the default but it was going to involve changing a bunch of tests, etc. so I decided to leave it alone and just fix the comment. For what it's worth, a logic for the current behavior is that if you think of the attributes of the child-most collection as the ungrouped attributes, adding a new collection is like pulling an attribute out of the ungrouped attributes and the most logical place to put that new collection is immediately adjacent to the child-most collection.